### PR TITLE
clusterlogforwarder adjustments + unit test added

### PR DESCRIPTION
**Pull request by elenagerman**
clusterlogforwarder adjustments + unit test added

### CODE REVIEW

File: pkg/clusterlogging/clusterlogforwarder.go [hunk 1]
```diff
@@ -35,6 +35,10 @@ func NewClusterLogForwarderBuilder(
 	builder := &ClusterLogForwarderBuilder{
 		apiClient: apiClient,
 		Definition: &clov1.ClusterLogForwarder{
+			TypeMeta: metav1.TypeMeta{
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder.go:38]: You can safely remove it.

File: pkg/clusterlogging/clusterlogforwarder_test.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+		{
+			name:          defaultClusterLogForwarderName,
+			namespace:     "",
+			expectedError: "clusterlogforwarder 'nsname' cannot be empty",
+		},
+	}
+
+	for _, testCase := range testCases {
+		testSettings := clients.GetTestClients(clients.TestClientParams{
+			GVK: []schema.GroupVersionKind{clusterLogForwarderGVK},
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder_test.go:149]: You can remove it

File: pkg/clusterlogging/clusterlogforwarder_test.go [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+		apiClient, "", defaultClusterLogForwarderNamespace)
+	clusterLogForwarderBuilder.Definition.ResourceVersion = "999"
+
+	return clusterLogForwarderBuilder
+}
+
+func buildClusterLogForwarderClientWithDummyObject() *clients.Settings {
+	return clients.GetTestClients(clients.TestClientParams{
+		K8sMockObjects: buildDummyClusterLogForwarder(),
+		GVK:            []schema.GroupVersionKind{clusterLogForwarderGVK},
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder_test.go:404]: You can safely remove it.

**Review by kononovn — CHANGES_REQUESTED** [review]
Looks good. Please remove few unnecessary lines

File: pkg/clusterlogging/clusterlogforwarder_test.go (hunk 2 above)
Comment (elenagerman) [reply pkg/clusterlogging/clusterlogforwarder_test.go:149]: I cannot remove it:
```go
=== RUN   TestClusterLogForwarderGet
    clusterlogforwarder_test.go:210: 
                Error Trace:    /home/elgerman/eco-gotest/eco-goinfra-ir/pkg/clusterlogging/clusterlogforwarder_test.go:210
                Error:          Not equal: 
                                expected: &v1.ClusterLogForwarder{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"instance", GenerateName:"", Namespace:"openshift-logging", SelfLink:"", UID:"", ResourceVersion:"999", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.ClusterLogForwarderSpec{Inputs:[]v1.InputSpec(nil), Outputs:[]v1.OutputSpec(nil), Filters:[]v1.FilterSpec(nil), Pipelines:[]v1.PipelineSpec(nil), ServiceAccountName:"", OutputDefaults:(*v1.OutputDefaults)(nil)}, Status:v1.ClusterLogForwarderStatus{Conditions:status.Conditions(nil), Inputs:v1.NamedConditions(nil), Outputs:v1.NamedConditions(nil), Filters:v1.NamedConditions(nil), Pipelines:v1.NamedConditions(nil)}}
                                actual  : &v1.ClusterLogForwarder{TypeMeta:v1.TypeMeta{Kind:"ClusterLogForwarder", APIVersion:"logging.openshift.io/v1"}, ObjectMeta:v1.ObjectMeta{Name:"instance", GenerateName:"", Namespace:"openshift-logging", SelfLink:"", UID:"", ResourceVersion:"999", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.ClusterLogForwarderSpec{Inputs:[]v1.InputSpec(nil), Outputs:[]v1.OutputSpec(nil), Filters:[]v1.FilterSpec(nil), Pipelines:[]v1.PipelineSpec(nil), ServiceAccountName:"", OutputDefaults:(*v1.OutputDefaults)(nil)}, Status:v1.ClusterLogForwarderStatus{Conditions:status.Conditions(nil), Inputs:v1.NamedConditions(nil), Outputs:v1.NamedConditions(nil), Filters:v1.NamedConditions(nil), Pipelines:v1.NamedConditions(nil)}}

```

File: pkg/clusterlogging/clusterlogforwarder.go (hunk 1 above)
Comment (elenagerman) [reply pkg/clusterlogging/clusterlogforwarder.go:38]: now all is broken on missing GVK if I remove it

Comment (elenagerman) [reply pkg/clusterlogging/clusterlogforwarder.go:38]: === RUN   TestClusterLogForwarderGet
    clusterlogforwarder_test.go:210: 
                Error Trace:    /home/elgerman/eco-gotest/eco-goinfra-ir/pkg/clusterlogging/clusterlogforwarder_test.go:210
                Error:          Not equal: 
                                expected: &v1.ClusterLogForwarder{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"instance", GenerateName:"", Namespace:"openshift-logging", SelfLink:"", UID:"", ResourceVersion:"999", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.ClusterLogForwarderSpec{Inputs:[]v1.InputSpec(nil), Outputs:[]v1.OutputSpec(nil), Filters:[]v1.FilterSpec(nil), Pipelines:[]v1.PipelineSpec(nil), ServiceAccountName:"", OutputDefaults:(*v1.OutputDefaults)(nil)}, Status:v1.ClusterLogForwarderStatus{Conditions:status.Conditions(nil), Inputs:v1.NamedConditions(nil), Outputs:v1.NamedConditions(nil), Filters:v1.NamedConditions(nil), Pipelines:v1.NamedConditions(nil)}}
                                actual  : &v1.ClusterLogForwarder{TypeMeta:v1.TypeMeta{Kind:"ClusterLogForwarder", APIVersion:"logging.openshift.io/v1"}, ObjectMeta:v1.ObjectMeta{Name:"instance", GenerateName:"", Namespace:"openshift-logging", SelfLink:"", UID:"", ResourceVersion:"999", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.ClusterLogForwarderSpec{Inputs:[]v1.InputSpec(nil), Outputs:[]v1.OutputSpec(nil), Filters:[]v1.FilterSpec(nil), Pipelines:[]v1.PipelineSpec(nil), ServiceAccountName:"", OutputDefaults:(*v1.OutputDefaults)(nil)}, Status:v1.ClusterLogForwarderStatus{Conditions:status.Conditions(nil), Inputs:v1.NamedConditions(nil), Outputs:v1.NamedConditions(nil), Filters:v1.NamedConditions(nil), Pipelines:v1.NamedConditions(nil)}}

File: pkg/clusterlogging/clusterlogforwarder_test.go [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+			assert.Equal(t, []clov1.PipelineSpec{*testCase.testPipeline}, result.Definition.Spec.Pipelines)
+		}
+	}
+}
+
+func buildValidClusterLogForwarderBuilder(apiClient *clients.Settings) *ClusterLogForwarderBuilder {
+	clusterLogForwarderBuilder := NewClusterLogForwarderBuilder(
+		apiClient, defaultClusterLogForwarderName, defaultClusterLogForwarderNamespace)
+	clusterLogForwarderBuilder.Definition.ResourceVersion = "999"
+	clusterLogForwarderBuilder.Definition.TypeMeta = metav1.TypeMeta{
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder_test.go:381]: You can safely remove it

File: pkg/clusterlogging/clusterlogforwarder_test.go [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+	}
+
+	return clusterLogForwarderBuilder
+}
+
+func buildInValidClusterLogForwarderBuilder(apiClient *clients.Settings) *ClusterLogForwarderBuilder {
+	clusterLogForwarderBuilder := NewClusterLogForwarderBuilder(
+		apiClient, "", defaultClusterLogForwarderNamespace)
+	clusterLogForwarderBuilder.Definition.ResourceVersion = "999"
+	clusterLogForwarderBuilder.Definition.TypeMeta = metav1.TypeMeta{
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder_test.go:393]: You can safely remove it

**Review by kononovn — CHANGES_REQUESTED** [review]
Please try to remove TypeMeta. Since this package has perfect integration in our client I think you don't need to use TypeMeta.

File: pkg/clusterlogging/clusterlogforwarder_test.go (hunk 4 above)
Comment (elenagerman) [reply pkg/clusterlogging/clusterlogforwarder_test.go:381]: I can not
```shell
=== RUN   TestClusterLogForwarderCreate
    clusterlogforwarder_test.go:229: 
                Error Trace:    /home/elgerman/eco-gotest/eco-goinfra-ir/pkg/clusterlogging/clusterlogforwarder_test.go:229
                Error:          Not equal: 
                                expected: &v1.ClusterLogForwarder{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"instance", GenerateName:"", Namespace:"openshift-logging", SelfLink:"", UID:"", ResourceVersion:"999", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.ClusterLogForwarderSpec{Inputs:[]v1.InputSpec(nil), Outputs:[]v1.OutputSpec(nil), Filters:[]v1.FilterSpec(nil), Pipelines:[]v1.PipelineSpec(nil), ServiceAccountName:"", OutputDefaults:(*v1.OutputDefaults)(nil)}, Status:v1.ClusterLogForwarderStatus{Conditions:status.Conditions(nil), Inputs:v1.NamedConditions(nil), Outputs:v1.NamedConditions(nil), Filters:v1.NamedConditions(nil), Pipelines:v1.NamedConditions(nil)}}
                                actual  : &v1.ClusterLogForwarder{TypeMeta:v1.TypeMeta{Kind:"ClusterLogForwarder", APIVersion:"logging.openshift.io/v1"}, ObjectMeta:v1.ObjectMeta{Name:"instance", GenerateName:"", Namespace:"openshift-logging", SelfLink:"", UID:"", ResourceVersion:"999", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.ClusterLogForwarderSpec{Inputs:[]v1.InputSpec(nil), Outputs:[]v1.OutputSpec(nil), Filters:[]v1.FilterSpec(nil), Pipelines:[]v1.PipelineSpec(nil), ServiceAccountName:"", OutputDefaults:(*v1.OutputDefaults)(nil)}, Status:v1.ClusterLogForwarderStatus{Conditions:status.Conditions(nil), Inputs:v1.NamedConditions(nil), Outputs:v1.NamedConditions(nil), Filters:v1.NamedConditions(nil), Pipelines:v1.NamedConditions(nil)}}
                                
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -2,4 +2,4 @@
                                  TypeMeta: (v1.TypeMeta) {
                                -  Kind: (string) "",
                                -  APIVersion: (string) ""
                                +  Kind: (string) (len=19) "ClusterLogForwarder",
                                +  APIVersion: (string) (len=23) "logging.openshift.io/v1"
                                  },
                Test:           TestClusterLogForwarderCreate

```

Comment (elenagerman) [reply pkg/clusterlogging/clusterlogforwarder_test.go:381]: this was a WA do not fail after removing you ask me before

Comment (elenagerman) [reply pkg/clusterlogging/clusterlogforwarder_test.go:381]: @kononovn I see how you solve it for the clusterlogging, but why to compare just name and namespace if we can compare the object itself?

File: pkg/clusterlogging/const.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package clusterlogging
+
+const (
```
Comment (kononovn) [inline pkg/clusterlogging/const.go:3]: Please remove some unused variables and move rest of them to _test file + make them private.

File: pkg/clusterlogging/clusterlogforwarder.go [hunk 7]
```diff
@@ -114,36 +114,42 @@ func (builder *ClusterLogForwarderBuilder) WithPipeline(pipelineSpec *clov1.Pipe
 }
 
 // PullClusterLogForwarder retrieves an existing clusterlogforwarder object from the cluster.
-func PullClusterLogForwarder(apiClient *clients.Settings, name, namespace string) (*clov1.ClusterLogForwarder, error) {
-	glog.V(100).Infof("Pulling existing clusterlogforwarder %s in namespace %s", name, namespace)
+func PullClusterLogForwarder(apiClient *clients.Settings, name, nsname string) (*ClusterLogForwarderBuilder, error) {
+	if apiClient == nil {
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder.go:118]: Please move this verification after log message:
glog.V(100).Infof("Pulling existing clusterlogforwarder %s in nsname %s", name, nsname)

File: pkg/clusterlogging/clusterlogforwarder_test.go [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+func TestClusterLogForwarderCreate(t *testing.T) {
+	testCases := []struct {
+		testClusterLogForwarder *ClusterLogForwarderBuilder
+		expectedError           string
+	}{
+		{
+			testClusterLogForwarder: buildValidClusterLogForwarderBuilder(buildClusterLogForwarderClientWithDummyObject()),
+			expectedError:           "",
+		},
+		{
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder_test.go:218]: Please create 1 more test case:
{
testClusterLogForwarder: buildValidClusterLogForwarderBuilder(clients.GetTestClients(clients.TestClientParams{})),
expectedError:           "",
		},

File: pkg/clusterlogging/clusterlogforwarder_test.go [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+func TestClusterLogForwarderDelete(t *testing.T) {
+	testCases := []struct {
+		testClusterLogForwarder *ClusterLogForwarderBuilder
+		expectedError           error
+	}{
+		{
+			testClusterLogForwarder: buildValidClusterLogForwarderBuilder(buildClusterLogForwarderClientWithDummyObject()),
+			expectedError:           nil,
+		},
+		{
```
Comment (kononovn) [inline pkg/clusterlogging/clusterlogforwarder_test.go:245]: Please create 1 more test case:
{
testClusterLogForwarder: buildValidClusterLogForwarderBuilder(clients.GetTestClients(clients.TestClientParams{})),
expectedError: "",
},

**Review by kononovn — CHANGES_REQUESTED** [review]
Ok looks good. 3 more comments and lgmt

**Review by kononovn — APPROVED** [review]
/lgtm

**Pull request merged**